employee-rostering-server/src/test/java/org/optaweb/employeerostering/server/solver/SolverTest.java

```diff
@@ -180,4 +544,153 @@ protected RosterGenerator buildRosterGenerator() {
+                              Integer yearlyHours, Duration timeBetweenShifts, Duration periodLength) {
+            this.constraint = constraint;
+            this.dailyHours = dailyHours;
+            this.weeklyHours = weeklyHours;
+            this.monthlyHours = monthlyHours;
+            this.yearlyHours = yearlyHours;
+            this.timeBetweenShifts = timeBetweenShifts;
+            this.periodLength = periodLength;
+        }
+
+        public Constraints getConstraint() {
+            return constraint;
+        }
+
+        public Contract getContract(AtomicLong idGenerator) {
+            Contract out = new Contract(TENANT_ID, "Contract", dailyHours, weeklyHours, monthlyHours,
+                                        yearlyHours);
+            out.setId(idGenerator.getAndIncrement());
+            return out;
+        }
+
+        public List<Shift> generateShifts(ShiftBuilder shiftBuilder) {
+            List<Shift> out = new ArrayList<>();
+            shiftBuilder.withTimeBetweenShifts(timeBetweenShifts);
+            out.addAll(shiftBuilder.generateShifts(3));
+            shiftBuilder.startingAtDate(shiftBuilder.firstShiftStartTime.plus(periodLength));
+            out.addAll(shiftBuilder.generateShifts(3));
+            return out;
+        }
+    }
+
+    private static class ShiftBuilder {
+
+        OffsetDateTime firstShiftStartTime;
+        Duration lengthOfShift;
+        Duration durationBetweenShifts;
+        Spot shiftSpot;
+        AtomicLong idGenerator;
+
+        public ShiftBuilder(AtomicLong idGenerator) {
+            this.idGenerator = idGenerator;
+        }
+
+        public ShiftBuilder startingAtDate(OffsetDateTime startDate) {
+            this.firstShiftStartTime = startDate;
+            return this;
+        }
+
+        public ShiftBuilder withShiftLength(Duration duration) {
+            this.lengthOfShift = duration;
+            return this;
+        }
+
+        public ShiftBuilder withTimeBetweenShifts(Duration duration) {
+            this.durationBetweenShifts = duration;
+            return this;
+        }
+
+        public ShiftBuilder forSpot(Spot spot) {
+            this.shiftSpot = spot;
+            return this;
+        }
+
+        public List<Shift> generateShifts(int numberOfShifts) {
+            if (firstShiftStartTime == null || lengthOfShift == null || durationBetweenShifts == null || shiftSpot == null) {
+                throw new IllegalStateException("ShiftBuilder not initialized");
+            }
+            List<Shift> out = new ArrayList<>();
+            OffsetDateTime shiftStart = firstShiftStartTime;
+
+            for (int i = 0; i < numberOfShifts; i++, shiftStart = shiftStart.plus(durationBetweenShifts)) {
+                out.add(new Shift(TENANT_ID, shiftSpot, shiftStart, shiftStart.plus(lengthOfShift)));
+            }
+            out.forEach(s -> s.setId(idGenerator.getAndIncrement()));
```

Comment:
I think you can merge both `for` loops, if you access last element of `out` via `i` provided in the first loop.
Something like `out.get(i).setId(idGenerator.getAndIncrement());`